### PR TITLE
Refactor code and add support for dotbot defaults

### LINKED ISSUE #19: Use "brew bundle --verbose" for brewfile directives
`brew bundle` shows prettified output with green colors after installing a package and no styling in case of the package being already installed, so it looks great.

`brew bundle --verbose` shows more details on what is actually going on, in particular on whether a particular package is being installed. 

the non-verbose version is prettier but it looks like it's stuck or something on large installs. It doesn't show enough information.

Consider changing this?

### PR DESCRIPTION
**Pull request by miguelandres**
This PR adds support for defaults to each of the directives supported by
dotbot-brew. The defaults are booleans for 'stdin', 'stdout', 'stderr'
which control whether the underlying commands get any of these handles
or are redirected to /dev/null.

Furthermore this PR refactors code to improve code reuse throught the
whole Brew class and makes Brewfile use `brew bundle --verbose` and
adds a `force_intel` option to defaults for M1 macs.

Home-brew is still not supported natively in M1 macs, so they recommend
using the intel version for the time being

Fixes #19

### CODE REVIEW

**Comment by d12frosted** [comment]
Hey @miguelandres 

I need some time to review this PR, because it touches many unrelated aspects. Do you mind if I will gradually create PRs from your changes? You will stay as author of commit :)

**Comment by miguelandres** [comment]
Sure thing!   
  
  
  
  
  
 —Miguel Andrés     
  

  
  
>

File: README.org [hunk 1]
```diff
@@ -50,3 +62,30 @@ installed using =brew=. The same works with =cask= and =brewfile=. For example:
 
 - cask: [colorpicker, vlc]
 #+END_SRC
+
+Defaults also can control stdin, stdout and stderr (disabling with false, enabling with true),
+see below for an example
+
+#+BEGIN_SRC yaml
+- defaults:
+    - brewfile:
```
Comment (d12frosted) [inline README.org:71]: The way it's written, I would think that these configurations for `stdin`, `stdout` and `stderr` are applied only to `brewfile` directive. But as far as I understand from your code, it is applied to all commands because `_defaults` is a mutable variable (read other comments). Doesn't sound cool to me. As suggestion, this could be renamed to `dotbot-brew` to clearly indicate that this section configures behaviour of `dotbot-brew` as a whole and not just one of the directives it provides. Or defaults should passed as arguments (see my other comments).

File: brew.py [hunk 2]
```diff
@@ -1,104 +1,126 @@
...
+import os
+import platform
+import subprocess
+import dotbot
 class Brew(dotbot.Plugin):
     _brewDirective = "brew"
     _caskDirective = "cask"
     _tapDirective = "tap"
     _brewFileDirective = "brewfile"
+    _defaults = {}
```
Comment (d12frosted) [inline brew.py:10]: I don't like these two variables. There is no reason for them to be mutable. Would be nicer to get defaults and pass them to directives as an argument (e.g. `self._brew(data, defaults)`). This way it's easier to read the code and there are no unintended side effects.

File: brew.py [hunk 3]
```diff
@@ -1,104 +1,126 @@
...
+    _forceIntelOption = "force_intel"
```
Comment (d12frosted) [inline brew.py:11]: Same goes here.

File: brew.py [hunk 4]
```diff
@@ -1,104 +1,126 @@
...
 
     def can_handle(self, directive):
         return directive in (self._tapDirective, self._brewDirective, self._caskDirective, self._brewFileDirective)
 
     def handle(self, directive, data):
+        self._defaults = self._context.defaults().get(directive, {})
```
Comment (d12frosted) [inline brew.py:17]: Seems like here you wanted to get defaults for a given directive (e.g. `brewflie`, `tap`, etc.). But due to the fact that `self._defaults` has wider scope, it has a side effect of leaking these defaults to other directives.

File: brew.py [hunk 5]
```diff
@@ -1,104 +1,126 @@
...
+        return subprocess.call(cmd, shell=True, cwd=self._context.base_directory()) == 0
 
     def _bootstrap_cask(self):
-        self._bootstrap_brew()
-        cmd = "brew tap caskroom/cask"
-        self._bootstrap(cmd)
+        self._log.info("Installing cask")
+        cmd = "brew tap homebrew/cask"
+        if self._defaults.get(self._forceIntelOption, False) == True:
+            cmd = "arch --x86_64 " + cmd
```
Comment (d12frosted) [inline brew.py:125]: I am not sure how this works. The final command would be `arch --x86_64 brew tap homebrew/cask`. Could you please point me where I can read about the issue with ARM? I tried grepping `man brew`, but didn't find it.

File: .github/workflows/macos.yml [hunk 6]
```diff
@@ -14,5 +14,25 @@ jobs:
 
     steps:
       - uses: actions/checkout@v2
+
       - name: Install some packages
         run: ./test
+
+      - name: Test installed brew `ack`
```
Comment (d12frosted) [inline .github/workflows/macos.yml:21]: this is already in master

File: README.org [hunk 7]
```diff
@@ -3,10 +3,22 @@
 :ID:                     52d1c964-770e-4844-8ab4-2f7f395d97c7
 :END:
 
-Plugin for [[https://github.com/anishathalye/dotbot][dotbot]] that adds =brew= and =cask= directives. It allows installation of
-packages using =brew= and =brew cask= on OS X. In case =brew= is not installed, it
-will be automatically loaded and configured. The plugin itself is pretty silly
-as it doesn't handle updates and fails on unsupported operating systems.
+#+CAPTION: Integration test status
+#+NAME:   fig:INT-TESTS
```
Comment (d12frosted) [inline README.org:7]: this is already in master

**Review by d12frosted — CHANGES_REQUESTED** [review]
Thank you for resurrecting your work on `dotbot-brew`. I have reviewed
your changes and I still have questions. Due to the fact that your PR
changes many unrelated aspects, it is very hard for me to review it,
especially when I have questions to first commits. Checkout the diff
of your PR, it even contains changes that are already in master.

Unfortunately, I am very limited in time I can spend on this project,
so would really appreciate if you keep your PRs limited to 1 specific
issue/feature. I understand that you have many ideas (and I love
them), but seriously, it is very hard for me to process in the current
state.

And please keep the history clean, this is hard to follow (you have
repeating commits and  a merge commit):

```
*   0e20013 <Miguel Barreto> - (refs/pullreqs/28) Merge branch 'add_defaults' of github.com:miguelandres/dotbot-brew into add_defaults (8 days ago)
|\
| * 17b94d5 <Miguel Barreto> - use correct code syntax for org mode (8 days ago)
| * 6d01450 <Miguel Barreto> - describe force_intel in readme (8 days ago)
| * 9b3e102 <Miguel Barreto> - Add force_intel option to bootstrap step and pre-package-install check (8 days ago)
| * cc3b039 <Miguel Barreto> - Refactor code and add support for dotbot defaults (8 days ago)
* | 55e8c72 <Miguel Barreto> - use correct code syntax for org mode (8 days ago)
* | 6210fa0 <Miguel Barreto> - describe force_intel in readme (8 days ago)
* | 1d3f0e5 <Miguel Barreto> - Add force_intel option to bootstrap step and pre-package-install check (8 days ago)
* | 1ac8e88 <Miguel Barreto> - Refactor code and add support for dotbot defaults (8 days ago)
* | 1f82030 <Miguel Barreto> - (HEAD -> master, origin/master, origin/HEAD) (ci) improve testing infrastructure (8 days ago)
|/
* ff00c80 <Boris Buliga> - (origin/ci/periodical, refs/pullreqs/26, ci/periodical) (ci) run tests periodically (3 weeks ago)
```

The way I see how to split your work (including stuff from #22):

1. Add support for `dotobot` defaults with the following valid
   configuration values: `stdin`, `stdout` and `stderr` (in truth, I
   am not sure why people would want to configure `stdin`). This
   includes code changes and documentation for this feature.
2. Add `force_intel` configuration. This includes code changes,
   documentation for this option and preferably tests.
3. Improvement to cask bootstrap flow (so it doesn't take too much
   time).
   
Thanks,

I hope that makes sense to you and you understand me. Please let me
know if I can help you somehow.

**Comment by d12frosted** [comment]
Closing for inactivity.
